### DISCUSSION TITLE
Create Webviews at runtime

### DISCUSSION THREAD

**Discussion by Quentin-Tran** [Voltstro-Studios/UnityWebBrowser#177] [Q&A] [answered]
Hi !

I just discovered your package and it's very cool, thank you for sharing it !

However, I have a question, in my current project, I would like to instantiate webviews at runtime by code (for instance WebBrowserBasic). But each webview needs a CommunicationLayer with unique ports and it seems impossible to change the CommunicationLayer at runtime as WebBrowserClient.Init() is an internal method only called inside BaseUwbClientManager.Start(). So for now, I created a few WebBrowserBasic prefabs with unique CommunicationLayer but i'm limited by the number of prefabs created in advance.

Is there a solution to create webview directly at runtime ?

Thanks in advance !

**Comment by Voltstro** [comment, marked as answer]
You could just set [`WebBrowserClient.communicationLayer`](https://projects.voltstro.dev/UnityWebBrowser/api/voltstrostudios.unitywebbrowser.core/voltstrostudios.unitywebbrowser.core.webbrowserclient.communicationlayer/) to a new dynamically created CommunicationLayer instance. You will probably need to do this in Unity's Awake method.

**Reply by Quentin-Tran** [reply]
Thank you, I was too focused on trying to init the browser later, I didn't think of using the Awake method !

**Comment by jacobbullock** [comment]
I am also trying to create this view at runtime, but not from an Awake method.  How should I go about setting the engine and communication layer.  I am just doing something like 

```
GameObject object = new GameObject("Browser");
WebBrowserUIBasic browser = object.AddComponent<WebBrowserUIBasic>();
BaseUwbClientManager manager = object.GetComponent<BaseUwbClientManager>();
WebBrowserClient. client = manager.browserClient;
client.javascript = true;
client.engine = ???;
```

**Reply by Voltstro** [reply]
Both engine and communication layer are scriptable objects. You will need to either:
- Store a reference to a set scriptable object (then set the variable to that).
- If you want to create them dynamically too, use Unity's [`ScriptableObject.CreateInstance`](https://docs.unity3d.com/2021.3/Documentation/ScriptReference/ScriptableObject.CreateInstance.html) method to create one. You will need to set the values.

**Reply by jacobbullock** [reply]
@Voltstro I appear to have this working to load dynamically at least a little.  However, right now, I don't seem to be able to focus the text fields or enter text in the site loaded.  Is there anything specific that needs to happen?

this is how I am creating the web view dynamically. 

```
windows_webView_container = new GameObject("WebViewObjectContainer");
            Canvas canvas = windows_webView_container.AddComponent<Canvas>();
            CanvasScaler canvasScaler = windows_webView_container.AddComponent<CanvasScaler>();

            windows_webView = new GameObject("WebViewObject");
            windows_webView.transform.parent = windows_webView_container.transform;

            WebBrowserUIBasic browser = windows_webView.AddComponent<WebBrowserUIBasic>();
            clientmanager = windows_webView.GetComponent<BaseUwbClientManager>();

            browser.inputHandler = ScriptableObject.CreateInstance<WebBrowserOldInputHandler>();
            webClient = clientmanager.browserClient;
            webClient.jsMethodManager.jsMethodsEnable = true;
            webClient.javascript = true;
            EngineConfiguration config = ScriptableObject.CreateInstance<EngineConfiguration>();
            config.engineAppName = "UnityWebBrowser.Engine.Cef";
            EnginePlatformFiles file = new Engine.EnginePlatformFiles();
            file.platform = Platform.Windows64;
            file.engineFileLocation = "Packages/dev.voltstro.unitywebbrowser.engine.cef.win.x64/Engine~/";
            config.engineFiles = new EnginePlatformFiles[] { file };
            webClient.engine = config;
            webClient.communicationLayer = ScriptableObject.CreateInstance<TCPCommunicationLayer>();
            webClient.RegisterJsMethod<string>("_cb", _cb);```
            
            
            the page is rendering, although I still need to finesse some things because it renders upside down and backwards by default. But even changing those settings in runtime, the text input never works.

**Reply by Voltstro** [reply]
I've added a [`UWBRuntime`](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/master/src/UnityWebBrowser.UnityProject/Assets/Scripts/UWBRuntime.cs) example script, that creates the canvas, the raw image and UWB (as well as anything else required) dynamically. Its renders correctly and doesn't have any input issues.

(You can also import the sample script from the samples section on the UWB package in UPM with the latest 2.1.1 release)

**Comment by jacobbullock** [comment]
@Voltstro this gist worked perfectly for us and filled in the rest of the gaps I was dealing with.  Thanks again for all your help.
